orderer/common/cluster/block_fetcher.go

```diff
@@ -0,0 +1,503 @@
+	}
+
+	// evict one entry if s.cache is full upto s.max
+	if uint(len(s.cache))+1 > s.max && s.max > 0 {
+		s.evict()
+	}
+
+	s.lastIndex++
+	s.cache[entry] = s.lastIndex
+}
+
+func (s *suspectSet) evict() {
+
+	keys := make([]string, 0, len(s.cache))
+
+	for k := range s.cache {
+		keys = append(keys, k)
+	}
+	// sort keys by ascending order of values
+	sort.SliceStable(keys, func(i, j int) bool {
+		return s.cache[keys[i]] < s.cache[keys[j]]
+	})
+
+	// remove the first element of keys
+	s.remove(keys[0])
+}
+
+func (s *suspectSet) remove(entry string) {
+	delete(s.cache, entry)
+}
+
+func (s suspectSet) has(entry string) bool {
+	_, ok := s.cache[entry]
+	return ok
+}
+
+type groupCollection struct {
+	elems []EndpointCriteria
+}
+
+func (g groupCollection) getIdxForEndpoint(en string) (int, error) {
+	var err error
+	for idx, e := range g.elems {
+		if e.Endpoint == en {
+			return idx, err
+		}
+	}
+
+	return 0, errors.New("endpoint not in the list")
+}
+
+//neighbours returns n neighbours of elem form the groupCollection
+func (g groupCollection) neighbours(elem EndpointCriteria, n uint64) ([]EndpointCriteria, error) {
+	currIdx, err := g.getIdxForEndpoint(elem.Endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	nbr := []EndpointCriteria{}
+	for i := uint64(currIdx + 1); i <= (n + uint64(currIdx)); i++ {
+		nbr = append(nbr, g.elems[i%(uint64(len(g.elems)))])
+	}
+	return nbr, nil
+}
+
+type AttestationPuller struct {
+	Config FetcherConfig
+	// Internal state
+	openConnections []attestationConnectionInfo
+	cancelStream    func()
+	Logger          *flogging.FabricLogger
+}
+
+func (p *AttestationPuller) seekNextEnvelope(startSeq uint64) (*common.Envelope, error) {
+	return protoutil.CreateSignedEnvelopeWithTLSBinding(
+		common.HeaderType_DELIVER_SEEK_INFO,
+		p.Config.Channel,
+		p.Config.Signer,
+		nextSeekInfo(startSeq),
+		int32(0),
+		uint64(0),
+		util.ComputeSHA256(p.Config.TLSCert),
+	)
+}
+
+func (p *AttestationPuller) closeEndpoints() {
+	// Close the streams and grpc connections
+	for _, oc := range p.openConnections {
+		oc.stream.cancelFunc()
+		if oc.conn != nil {
+			oc.conn.Close()
+		}
+	}
+	p.openConnections = nil
+}
+
+func (p *AttestationPuller) Close() {
+	p.closeEndpoints()
+}
+
+func (p *AttestationPuller) pullAttestationBlock(ec EndpointCriteria, seq uint64, env *common.Envelope) (*orderer.BlockAttestation, error) {
+
+	// The connection should be via closed via Close() method of AttestationPuller
+	conn, err := p.Config.Dialer.Dial(ec)
+	if err != nil {
+		return nil, err
+	}
+
+	stream, err := newImpatientStream(conn, p.Config.FetchTimeout, env)
+	if err != nil {
+		return nil, err
+	}
+
+	// append the stream to bf.openConnections, so that Close can close all streams
+	// dial connections
+	cni := attestationConnectionInfo{stream: stream, conn: conn}
+	p.openConnections = append(p.openConnections, cni)
+
+	resp, err := stream.Recv()
+	status := resp.GetStatus()
+	if status != common.Status_SUCCESS {
```

Comment:
it's either a status or a block attestation but it cannot be both :-)

Reply:
fixed